lib/IRGen/GenDecl.cpp

```diff
@@ -3697,12 +3727,60 @@ llvm::Constant *IRGenModule::emitSwiftProtocols() {
     break;
   }
 
-  var->setSection(sectionName);
-  
-  disableAddressSanitizer(*this, var);
-  
-  addUsedGlobal(var);
-  return var;
+  // For JIT, emit the protocol list as a single global array, and return it.
+  if (asContiguousArray) {
+    ConstantInitBuilder builder(*this);
+    auto recordsArray = builder.beginArray(ProtocolRecordTy);
+    for (auto *protocol : SwiftProtocols) {
+      auto record = recordsArray.beginStruct(ProtocolRecordTy);
+      // Relative reference to the protocol descriptor.
+      auto descriptorRef = getAddrOfLLVMVariableOrGOTEquivalent(
+          LinkEntity::forProtocolDescriptor(protocol));
+      record.addRelativeAddress(descriptorRef);
+      record.finishAndAddTo(recordsArray);
+    }
+
+    // Define the global variable for the protocol list.
+    // FIXME: This needs to be a linker-local symbol in order for Darwin ld to
+    // resolve relocations relative to it.
+    auto var = recordsArray.finishAndCreateGlobal(
+        "\x01l_protocols", Alignment(4),
+        /*isConstant*/ true, llvm::GlobalValue::PrivateLinkage);
+    var->setSection(sectionName);
+    disableAddressSanitizer(*this, var);
+    addUsedGlobal(var);
+
+    return var;
+  }
+
+  // In non-JIT mode, emit the protocol records as individual globals.
+  for (auto *protocol : SwiftProtocols) {
+    auto entity = LinkEntity::forProtocolDescriptor(protocol);
+    auto link = LinkInfo::get(*this, entity, NotForDefinition);
+    auto var = new llvm::GlobalVariable(
+        Module, ProtocolRecordTy, /*isConstant*/ true,
+        llvm::GlobalValue::PrivateLinkage, /*initializer*/ nullptr,
+        "\x01l_protocol_" + link.getName());
```

Comment:
That naming convention breaks the ability to demangle the symbol. Also, it adds a lot of symbol name string data.
I'd prefer to add a mangling scheme for these variables.